content/blog/2020-01-ggplot2-3-3-0/index.Rmarkdown

```diff
@@ -0,0 +1,231 @@
+
+```{r}
+ggplot(mpg) + 
+  geom_bar(aes(y = manufacturer))
+```
+
+The direction determination is not only looking at which aesthetics are mapped, but also what they are mapped to, so most layers can be determined without ambiguity. In the presence of ambiguity it is conservative and will default to the standard orientation. If you want to overwrite the direction, either because it fails to detect it, or because the geom/stat is ambiguous by nature, It can be set directly with the `orientation` argument:
+
+```{r, message=FALSE}
+ggplot(mpg, aes(displ, hwy)) + 
+  geom_point() + 
+  geom_smooth(orientation = "y")
+```
+
+### More control over aesthetic evaluation
+When mapping data to aesthetics we are usually differentiating between two different mappings. One for the stat and one for the geom. We don't think about this that often, because the stat simply accept the mappings and carry them over to the geom (sometimes with modifications). Sometimes, though, we want to control the mapping that happens between the output of the stat and the geom. The approach to this has been to use `stat()` (and `..var..` notation before that), to indicate that the aesthetic should be evaluated after the stat has been computed.
+
+With this release we expand on this, as well as making it more explicit. There are now three entry points to evaluation: *start*, *after_stat*, and *after_scale*. *start* is the default (i.e. evaluated in the context of the layer data), *after_stat* is like the old `stat()` (i.e. evaluated in the context of the stat output), and *after_scale* is new and specifies that evaluation should occur after the aesthetic has been scaled. The `after_stat()` and `after_scale()` functions are used to mark aesthetics for their respective delayed evaluation (thus soft-deprecating `stat()`):
+
+You would use `after_stat()` to e.g. pick a different computed metric than the default (here _density_ instead of _count_)
+
+```{r, message=FALSE}
+ggplot(mpg, aes(displ)) +
+  geom_histogram(aes(y = after_stat(density)))
+```
+
+And you can use `after_scale()` to assign fill as a variant of the scaled color:
+
+```{r}
+ggplot(mpg, aes(class, hwy)) +
+  geom_boxplot(aes(colour = class, fill = after_scale(alpha(colour, 0.4))))
+```
+
+Furthermore, it is now possible to perform multiple mapping for the same aesthetic, using the `stage()` function. This can be used to e.g. set alpha on the stroke as the default is to only apply alpha to fill for polygon-type geoms:
+
+```{r}
+ggplot(mpg) + 
+  geom_bar(
+    aes(
+      x = drv, 
+      colour = stage(start = drv, after_scale = alpha(colour, 0.5))
+    ), 
+    fill = NA, size = 4
+  )
+```
+
+`stage()` is definitely not something you need every day, but it does resolve a range of issues (like the one above) and it is nice to have full transparency and control over the aesthetic evaluation.
+
+### More extensible theming
```

Comment:
In this section, I think it would be helpful to also see an example of how code can then make use of the new theme element, just like I did in the examples to `register_theme_elements()`. Apart from making the example less abstract, it'll also show the use of `element_render()`. As much as possible, we don't want extension developers to work with the contents of the theme elements directly. We may even want to state that they should use `calc_element()` to access and `element_render()` to render the theme elements.

Reply:
The reason I didn't do this was that I didn't want to drag in the creation of a new coord etc. just for that. I think that's beyond the scope of a release post. If you really think it is necessary I can add it, though

Reply:
It's fine if you think it's beyond the scope. In this case, maybe refer the reader to the documentation of `register_theme_elements()` for a complete working example.

Reply:
good idea